siliconcompiler/core.py

```diff
@@ -563,40 +567,84 @@ def load_target(self, name):
 
         Args:
             name (str): Module name
-            flow (str): Target flow to
+            **kwargs (str): Options to pass along to the target
 
         Examples:
-            >>> chip.load_target('freepdk45_demo')
-            Loads the 'freepdk45_demo' target
-
+            >>> chip.load_target('freepdk45_demo', syn_np=5)
+            Loads the 'freepdk45_demo' target with 5 parallel synthesis tasks
         """
 
         # Record target
         self.set('option', 'target', name)
 
         load_function = self.find_function(name, 'setup', 'targets')
-        load_function(self)
+        load_function(self, **kwargs)
 
     ##########################################################################
-    def use(self, module):
+    def use(self, module, **kwargs):
         '''
         Loads a SiliconCompiler module into the current chip object by calling
         a module.setup() method.
         '''
 
+        # Load supported types here to avoid cyclic import
+        from siliconcompiler import PDK
+        from siliconcompiler import Flow
+        from siliconcompiler import Library
+        from siliconcompiler import Checklist
+
+        # Mark self as read only
+        self._readonly = True
+
         # Call the module setup function.
-        # TODO: Move from 'module.setup(self)' to 'self.[merge](module.setup())'
-        new_schema = module.setup(self)
-
-        # TODO: Is this the best way to determine module type? Also, remove chip arg from all but target?
-        if 'targets' in module.__name__:
-            self.set('option', 'target', module.__name__[module.__name__.rfind('.')+1:])
-        elif 'libs.' in module.__name__:
-            lib_name = module.__name__[module.__name__.rfind('.')+1:]
-            self._loaded_modules['libs'].append(lib_name)
-            # TODO: Stdcell libs may call 'import_library' of their own volition, and not return a Chip.
-            if new_schema:
-                self.import_library(module.setup(self))
+        use_modules = module.setup(self, **kwargs)
+
+        # Reset readonly flag
+        self._readonly = False
```

Comment:
Could the `PDK.setup` methods also call this to modify the parent `Chip`, if a package author got frustrated with the limitation?

Reply:
I'm not sure I follow about the `PDK.setup`?

Reply:
I mean, is there anything that makes the `_readonly` value immutable, or could a PDK module author call `chip._readonly = False` in the PDK `setup()` method, and then call `chip.set(...)`?

Reply:
Ahh, yes, for the same reason that anyone could do `chip.schema.cfg` that would only really be private if it was `self.__readonly`